engines/wintermute/base/gfx/opengl/x/meshx.cpp

```diff
@@ -0,0 +1,486 @@
+				int currentMaterialIndexTmp = lexer.tokenToInt();
+
+				// again, this assumes that face indices are only increasing
+				if (currentMaterialIndex < currentMaterialIndexTmp) {
+					currentMaterialIndex = currentMaterialIndexTmp;
+					_indexRanges.push_back(3 * i);
+				}
+
+				lexer.next();
+				lexer.next();
+			}
+
+			_indexRanges.push_back(3 * faceCount);
+			_numAttrs = _indexRanges.size() - 1;
+
+			while(!lexer.eof()) {
+				if (lexer.tokenIsIdentifier("Material")) {
+					lexer.next();
+					Material *mat = new Material(_gameRef);
+					mat->loadFromX(lexer, filename);
+					_materials.add(mat);
+				} else if (lexer.tokenIsIdentifier()) {
+					while (!lexer.reachedClosedBraces()) {
+						lexer.next();
+					}
+
+					lexer.next();
+				} else if (lexer.reachedClosedBraces()) {
+					break;
+				} else {
+					warning("MeshX::loadFromX unknown token encountered while loading materials");
+					break;
+				}
+			}
+
+			lexer.next();
+		} else if (lexer.tokenIsIdentifier("Material")) {
+			lexer.next();
+			Material *mat = new Material(_gameRef);
+			mat->loadFromX(lexer, filename);
+			_materials.add(mat);
+		} else if (lexer.tokenIsIdentifier("XSkinMeshHeader")) {
+			_skinnedMesh = true;
+			lexer.next();
+			lexer.advanceOnOpenBraces();
+
+			// care about skinned animatons later
+			int maxSkinWeightsPerVertex = lexer.tokenToInt();
+			lexer.next();
+			lexer.next();
+			int maxSkinWeightsPerFace = lexer.tokenToInt();
+			lexer.next();
+			lexer.next();
+			int boneCount = lexer.tokenToInt();
+			lexer.next();
+			lexer.next();
+			lexer.next();
+		} else if (lexer.tokenIsIdentifier("SkinWeights")) {
+			_skinnedMesh = true;
+			lexer.next();
+			lexer.advanceOnOpenBraces();
+
+			skinWeightsList.resize(skinWeightsList.size() + 1);
+			SkinWeights &currSkinWeights = skinWeightsList.back();
+
+			currSkinWeights._boneName = lexer.tokenToString();
+			lexer.next();
+			lexer.next();
+
+			int weightCount = lexer.tokenToInt();
+			currSkinWeights._vertexIndices.resize(weightCount);
+			currSkinWeights._vertexWeights.resize(weightCount);
+			lexer.next();
+			lexer.next();
+
+			for (int i = 0; i < weightCount; ++i) {
+				currSkinWeights._vertexIndices[i] = lexer.tokenToInt();
+				lexer.next();
+				lexer.next();
+			}
+
+			for (int i = 0; i < weightCount; ++i) {
+				currSkinWeights._vertexWeights[i] = lexer.tokenToFloat();
+				lexer.next();
+				lexer.next();
+			}
+
+			// TODO: check if this is the right format
+			for (int r = 0; r < 4; ++r) {
+				for (int c = 0; c < 4; ++c) {
+					currSkinWeights._offsetMatrix(c, r) = lexer.tokenToFloat();
+					lexer.next();
+					lexer.next();
+				}
+			}
+
+			lexer.next();
+			lexer.next();
+		} else if (lexer.tokenIsIdentifier()) {
+			while (!lexer.reachedClosedBraces()) {
+				lexer.next();
+			}
+
+			lexer.next();
+		} else if (lexer.reachedClosedBraces()) {
+			lexer.next();
+			return true;
+		} else {
+			warning("MeshX::loadFromX unknown token encounterd");

```

Comment:
Consider dumping some identifier for the unknown token.

Reply:
Done.